walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,59 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> requiredAttendees = request.getAttendees();
+    long meetingDuration = request.getDuration();
+    List<TimeRange> meetingQueryOptions = new ArrayList<>();
+
+    if (meetingDuration > TimeRange.WHOLE_DAY.duration() || meetingDuration <= 0) {
+      return meetingQueryOptions;
+    }
+    
+    List<TimeRange> eventsSortedByStartTime = new ArrayList<>();
+    List<TimeRange> eventsSortedByEndTime = new ArrayList<>();
+
+    for (Event event : events) {
+      // Add event to events lists if meeting request contains attendees of the event  
+      if (!Collections.disjoint(requiredAttendees, event.getAttendees())) {
+        eventsSortedByStartTime.add(event.getWhen());
+        eventsSortedByEndTime.add(event.getWhen());
+      }
+    }
+    Collections.sort(eventsSortedByStartTime, TimeRange.ORDER_BY_START);
+    Collections.sort(eventsSortedByEndTime, TimeRange.ORDER_BY_END);
```

Comment:
Is this list needed except for the last element? If not can the last event end time be determined in a more efficient way?

Reply:
I changed it to using Collection.max() instead. Just out of curiosity, would I be right in thinking that this would now take O(n) time since it goes through list and keeps track of a max element  through comparisons instead of O(n*log n) time since that is the average case for most sorting algorithms? Also, in terms of space complexity, using Collection.max() no longer takes up space for an list of n elements which would be O(n) space complexity, but rather just a single element?

Reply:
> O(n) time since it goes through list and keeps track of a max element through comparisons instead of O(n*log n) time since that is the average case for most sorting algorithms?

Exactly right.

> Also, in terms of space complexity, using Collection.max() no longer takes up space for an list of n elements which would be O(n) space complexity, but rather just a single element?

Yes though max does need some space for the comparator but that can be GCed after the method finishes.